File: app/src/main/java/com/wegielek/signalychinese/viewmodels/MainViewModel.java
package com.wegielek.signalychinese.viewmodels;

import android.annotation.SuppressLint;
import android.app.Application;
import android.graphics.Path;
import android.util.Log;

import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import com.google.mlkit.vision.digitalink.Ink;
import com.wegielek.signalychinese.database.Dictionary;
import com.wegielek.signalychinese.database.Radicals;
import com.wegielek.signalychinese.database.SearchHistory;
import com.wegielek.signalychinese.repository.DictionaryRepository;

import java.sql.Timestamp;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.concurrent.Executors;

public class MainViewModel extends AndroidViewModel {

    public MutableLiveData<Ink.Stroke.Builder> mStrokeBuilder = new MutableLiveData<>();
    public MutableLiveData<ArrayList<Ink.Stroke.Builder>> mStrokesHistory = new MutableLiveData<>();
    public MutableLiveData<Ink.Builder> mInkBuilder = new MutableLiveData<>();
    public MutableLiveData<Path> mCurrentVisibleStroke = new MutableLiveData<>();
    public MutableLiveData<ArrayList<Path>> mVisibleStrokesHistory = new MutableLiveData<>();
    public MutableLiveData<Integer> mCursorPosition = new MutableLiveData<>();
    public MutableLiveData<List<String>> mCharactersList = new MutableLiveData<>();
    public MutableLiveData<List<String>> mDictionaryResultsList = new MutableLiveData<>();
    public MutableLiveData<List<String[]>> mRadicalsList = new MutableLiveData<>();
    public MutableLiveData<Boolean> mIsRadicalChosen = new MutableLiveData<>();
    public MutableLiveData<String> mRadicalChosenCharacter = new MutableLiveData<>();
    private final DictionaryRepository mDictionaryRepository;

    public MainViewModel(Application application) {
        super(application);
        mDictionaryRepository = new DictionaryRepository(application);
        mStrokeBuilder.setValue(Ink.Stroke.builder());
        mInkBuilder.setValue(Ink.builder());
        mStrokesHistory.setValue(new ArrayList<>());
        mDictionaryResultsList.setValue(new ArrayList<>());
        mCharactersList.setValue(new ArrayList<>());
        mVisibleStrokesHistory.setValue(new ArrayList<>());
        mCurrentVisibleStroke.setValue(new Path());
        mCursorPosition.setValue(0);
        mRadicalChosenCharacter.setValue("0");
        mRadicalsList.setValue(new ArrayList<>());
        mIsRadicalChosen.setValue(false);
    }

    public LiveData<List<Radicals>> getRadicalsSection(String section) {
        setRadicalChosenCharacter(section);
        return mDictionaryRepository.getRadicalsSection(section);
    }

    public LiveData<List<Dictionary>> searchSingleCH(String searchQuery) {
        return mDictionaryRepository.searchSingleCH(searchQuery);
    }

    public LiveData<List<Dictionary>> searchByWordCH(String searchQuery) {
        return mDictionaryRepository.searchByWordCH(searchQuery);
    }

    public LiveData<List<Dictionary>> searchByWordPL(String searchQuery) {
        return mDictionaryRepository.searchByWordPL(searchQuery);
    }

    public String getRadicalChosenCharacter() {
        return mRadicalChosenCharacter.getValue();
    }

    public void setRadicalChosenCharacter(String radicalChosenCharacter) {
        mRadicalChosenCharacter.setValue(radicalChosenCharacter);
    }

    public String getResult(int index) {
        if (mDictionaryResultsList.getValue() != null) {
            String word = mDictionaryResultsList.getValue().get(index);

            @SuppressLint("SimpleDateFormat") SimpleDateFormat sdf =
                    new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
            Timestamp timestamp = new Timestamp(System.currentTimeMillis());
            String ts = sdf.format(timestamp);
            Log.i("TIME", ts);
            SearchHistory searchHistory = new SearchHistory();
            searchHistory.time = ts;
            searchHistory.traditionalSign = word.split("/")[0];
            searchHistory.simplifiedSign = word.split("/")[1];
            searchHistory.pronunciation = word.split("/")[2];
            searchHistory.pronunciationPhonetic = word.split("/")[3];
            int x = word.split("/")[0].length() + word.split("/")[1].length()
                    + word.split("/")[2].length() + word.split("/")[3].length() + 4;
            searchHistory.translation = word.substring(x);
            mDictionaryRepository.addHistoryRecord(searchHistory).addListener(() ->
                    Log.d("MainViewModel", "History record added successfully"), Executors.newSingleThreadExecutor()
            );

            return mDictionaryResultsList.getValue().get(index);
        } else {
            return null;
        }
    }

    public void setRadicalChosen(Boolean isRadicalChosen) {
        mIsRadicalChosen.setValue(isRadicalChosen);
    }

    public Boolean getRadicalChosen() {
        return mIsRadicalChosen.getValue();
    }

    public void updateResults(List<String> searchResults) {
        mDictionaryResultsList.setValue(searchResults);
    }

    public void addToCharacterList(String text) {
        if (mCharactersList.getValue() != null) {
            List<String> characterListCopy = new ArrayList<>(mCharactersList.getValue());
            characterListCopy.add(text);
            mCharactersList.setValue(characterListCopy);
        }
    }

    public void clearCharacterList() {
        mCharactersList.setValue(Collections.emptyList());
    }

    public void setCursorPosition(int x) {
        mCursorPosition.setValue(x);
    }

    public int getCursorPosition() {
        if (mCursorPosition.getValue() != null) {
            return mCursorPosition.getValue();
        } else {
            return -1;
        }
    }

    public void addVisibleStroke(Path path) {
        if (mVisibleStrokesHistory.getValue() != null) {
            ArrayList<Path> visibleStrokesHistoryCopy = new ArrayList<>(mVisibleStrokesHistory.getValue());
            visibleStrokesHistoryCopy.add(path);
            mVisibleStrokesHistory.setValue(visibleStrokesHistoryCopy);
        }
    }

    public void removeVisibleStroke(int index) {
        if (mVisibleStrokesHistory.getValue() != null) {
            ArrayList<Path> visibleStrokesHistoryCopy = new ArrayList<>(mVisibleStrokesHistory.getValue());
            visibleStrokesHistoryCopy.remove(index);
            mVisibleStrokesHistory.setValue(visibleStrokesHistoryCopy);
        }
    }

    public int getVisibleStrokeSize() {
        if (mVisibleStrokesHistory.getValue() != null) {
            return mVisibleStrokesHistory.getValue().size();
        } else {
            return -1;
        }
    }

    public Path getVisibleStroke(int index) {
        if (mVisibleStrokesHistory.getValue() != null) {
            return mVisibleStrokesHistory.getValue().get(index);
        } else {
            return null;
        }
    }

    public void clearVisibleStrokes() {
        mVisibleStrokesHistory.setValue(new ArrayList<>());
    }

    public Path getCurrentVisibleStroke() {
        return mCurrentVisibleStroke.getValue();
    }

    public void setCurrentVisibleStroke(Path path) {
        mCurrentVisibleStroke.setValue(path);
    }

    public void resetCurrentVisibleStroke() {
        mCurrentVisibleStroke.setValue(new Path());
    }

    public void lineToCurrentVisibleStroke(float x, float y) {
        Path currentVisibleStrokeCopy = new Path(mCurrentVisibleStroke.getValue());
        currentVisibleStrokeCopy.lineTo(x, y);
        mCurrentVisibleStroke.setValue(currentVisibleStrokeCopy);
    }

    public void moveToCurrentVisibleStroke(float x, float y) {
        Path currentVisibleStrokeCopy = new Path(mCurrentVisibleStroke.getValue());
        currentVisibleStrokeCopy.moveTo(x, y);
        mCurrentVisibleStroke.setValue(currentVisibleStrokeCopy);
    }

    public void setInkBuilder(Ink.Builder mInkBuilder) {
        this.mInkBuilder.setValue(mInkBuilder);
    }

    public Ink inkBuilderBuild() {
        if (mInkBuilder.getValue() != null) {
            return mInkBuilder.getValue().build();
        } else {
            return null;
        }
    }

    public void addInkBuilderStroke(Ink.Stroke stroke) {
        Ink.Builder builder = new Ink.Builder();
        Ink inkBuilderCopy = inkBuilderBuild();
        for (Ink.Stroke iStroke : inkBuilderCopy.getStrokes()) {
            builder.addStroke(iStroke);
        }
        builder.addStroke(stroke);
        mInkBuilder.setValue(builder);
    }

    public void removeFromStrokesHistory(int index) {
        if (mStrokesHistory.getValue() != null) {
            ArrayList<Ink.Stroke.Builder> strokesHistoryCopy = new ArrayList<>(mStrokesHistory.getValue());
            strokesHistoryCopy.remove(index);
            mStrokesHistory.setValue(strokesHistoryCopy);
        }
    }

    public int getStrokesHistorySize() {
        if (mStrokesHistory.getValue() != null) {
            return mStrokesHistory.getValue().size();
        } else {
            return -1;
        }
    }

    public void clearStrokesHistory() {
        mStrokesHistory.setValue(new ArrayList<>());
    }

    public void addToStrokesHistory(Ink.Stroke.Builder strokeBuilder) {
        if (mStrokesHistory.getValue() != null) {
            ArrayList<Ink.Stroke.Builder> strokesHistoryCopy = new ArrayList<>(mStrokesHistory.getValue());
            strokesHistoryCopy.add(strokeBuilder);
            mStrokesHistory.setValue(strokesHistoryCopy);
        }
    }

    public Ink.Stroke strokeHistoryBuild(int index) {
        if (mStrokesHistory.getValue() != null) {
            return mStrokesHistory.getValue().get(index).build();
        } else {
            return null;
        }
    }

    public Ink.Stroke.Builder getmStrokeBuilder() {
        return mStrokeBuilder.getValue();
    }

    public Ink.Stroke strokeBuilderBuild() {
        if (mStrokeBuilder.getValue() != null) {
            return mStrokeBuilder.getValue().build();
        } else {
            return null;
        }
    }

    public void addStrokeBuilderPoint(Ink.Point point) {
        Ink.Stroke.Builder builder = Ink.Stroke.builder();
        Ink.Stroke strokeBuilderCopy = strokeBuilderBuild();
        for (Ink.Point iPoint : strokeBuilderCopy.getPoints()) {
            builder.addPoint(iPoint);
        }
        builder.addPoint(point);
        mStrokeBuilder.setValue(builder);
    }

    public void clearStrokeBuilder() {
        mStrokeBuilder.setValue(Ink.Stroke.builder());
    }

    public void setRadicalsList(List<String[]> radicalsParentModels) {
        mRadicalsList.postValue(radicalsParentModels);
    }

}

File: app/src/main/java/com/wegielek/signalychinese/viewmodels/MainViewModel.kt
package com.wegielek.signalychinese.viewmodels

import android.annotation.SuppressLint
import android.app.Application
import android.graphics.Path
import android.util.Log
import androidx.core.content.ContextCompat
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.MutableLiveData
import com.google.common.util.concurrent.ListenableFuture
import com.google.mlkit.vision.digitalink.Ink
import com.wegielek.signalychinese.database.Dictionary
import com.wegielek.signalychinese.database.FlashCards
import com.wegielek.signalychinese.database.History
import com.wegielek.signalychinese.database.Radicals
import com.wegielek.signalychinese.repository.DictionaryRepository
import java.sql.Timestamp
import java.text.SimpleDateFormat

class MainViewModel(application: Application) : AndroidViewModel(application) {
    var charactersList = MutableLiveData<List<String>>()
    var dictionaryResultsList = MutableLiveData<List<Dictionary>>()
    var radicalsList = MutableLiveData<List<Array<String>>>()
    private val mStrokeBuilder = MutableLiveData<Ink.Stroke.Builder>()
    private val mStrokesHistory = MutableLiveData<ArrayList<Ink.Stroke.Builder>>()
    private val mInkBuilder = MutableLiveData<Ink.Builder>()
    private val mCurrentVisibleStroke = MutableLiveData<Path>()
    private val mVisibleStrokesHistory = MutableLiveData<ArrayList<Path>>()
    private val mCursorPosition = MutableLiveData<Int>()
    private val isRadicalChosen = MutableLiveData<Boolean>()
    private val mRadicalChosenCharacter = MutableLiveData<String>()
    private val mKeepSplashScreen = MutableLiveData<Boolean>()
    private val mDictionaryRepository: DictionaryRepository

    init {
        mDictionaryRepository = DictionaryRepository(application)
        mStrokeBuilder.value = Ink.Stroke.builder()
        mInkBuilder.value = Ink.builder()
        mStrokesHistory.value = ArrayList()
        dictionaryResultsList.value = ArrayList()
        charactersList.value = ArrayList()
        mVisibleStrokesHistory.value = ArrayList()
        mCurrentVisibleStroke.value = Path()
        mCursorPosition.value = 0
        mRadicalChosenCharacter.value = "0"
        radicalsList.value = ArrayList()
        isRadicalChosen.value = false
        mKeepSplashScreen.value = true
    }

    fun addFlashCardToGroup(flashCards: FlashCards) {
        mDictionaryRepository.addFlashCardToGroup(flashCards)
    }

    fun setKeepSplashScreen(keepSplashScreen: Boolean) {
        this.mKeepSplashScreen.value = keepSplashScreen
    }

    fun isKeepSplashScreen(): Boolean {
        return mKeepSplashScreen.value!!
    }

    fun getRadicalsSection(section: String): ListenableFuture<List<Radicals>> {
        setRadicalChosenCharacter(section)
        return mDictionaryRepository.getRadicalsSection(section)
    }

    fun searchSingleCH(searchQuery: String): ListenableFuture<List<Dictionary>> {
        return mDictionaryRepository.searchSingleCH(searchQuery)
    }

    fun searchByWordCH(searchQuery: String): ListenableFuture<List<Dictionary>> {
        return mDictionaryRepository.searchByWordCH(searchQuery)
    }

    fun searchByWordCHAll(searchQuery: String): ListenableFuture<List<Dictionary>> {
        return mDictionaryRepository.searchByWordCHAll(searchQuery)
    }

    fun searchByWordPL(searchQuery: String): ListenableFuture<List<Dictionary>> {
        return mDictionaryRepository.searchByWordPL(searchQuery)
    }

    fun searchByWordPLAll(searchQuery: String): ListenableFuture<List<Dictionary>> {
        return mDictionaryRepository.searchByWordPLAll(searchQuery)
    }

    val radicalChosenCharacter: String?
        get() = mRadicalChosenCharacter.value

    private fun setRadicalChosenCharacter(radicalChosenCharacter: String) {
        mRadicalChosenCharacter.value = radicalChosenCharacter
    }

    fun getResult(index: Int): Dictionary {
        val dictionary = dictionaryResultsList.value!![index]
        @SuppressLint("SimpleDateFormat") val sdf = SimpleDateFormat("yyyy-MM-dd HH:mm:ss")
        val timestamp = Timestamp(System.currentTimeMillis())
        val ts = sdf.format(timestamp)
        Log.i("TIME", ts)
        val history = History()
        history.time = ts
        history.traditionalSign = dictionary.traditionalSign
        history.simplifiedSign = dictionary.simplifiedSign
        history.pronunciation = dictionary.pronunciation
        history.pronunciationPhonetic = dictionary.pronunciationPhonetic
        history.translation = dictionary.translation
        mDictionaryRepository.addHistoryRecord(history).addListener(
            {
                Log.d(
                    "MainViewModel",
                    "History record added successfully"
                )
            },
            ContextCompat.getMainExecutor(getApplication<Application>().applicationContext)
        )
        return dictionaryResultsList.value!![index]
    }

    fun setRadicalChosen(isRadicalChosen: Boolean) {
        this.isRadicalChosen.value = isRadicalChosen
    }

    val radicalChosen: Boolean?
        get() = isRadicalChosen.value

    fun updateResults(searchResults: List<Dictionary>) {
        dictionaryResultsList.value = searchResults
    }

    fun addToCharacterList(text: String) {
        if (charactersList.value != null) {
            val characterListCopy = ArrayList(charactersList.value!!)
            characterListCopy.add(text)
            charactersList.value = characterListCopy
        }
    }

    fun clearCharacterList() {
        charactersList.value = emptyList()
    }

    var cursorPosition: Int
        get() = if (mCursorPosition.value != null) {
            mCursorPosition.value!!
        } else {
            -1
        }
        set(x) {
            mCursorPosition.value = x
        }

    fun addVisibleStroke(path: Path) {
        if (mVisibleStrokesHistory.value != null) {
            val visibleStrokesHistoryCopy = ArrayList(mVisibleStrokesHistory.value!!)
            visibleStrokesHistoryCopy.add(path)
            mVisibleStrokesHistory.value = visibleStrokesHistoryCopy
        }
    }

    fun removeVisibleStroke(index: Int) {
        if (mVisibleStrokesHistory.value != null) {
            val visibleStrokesHistoryCopy = ArrayList(mVisibleStrokesHistory.value!!)
            visibleStrokesHistoryCopy.removeAt(index)
            mVisibleStrokesHistory.value = visibleStrokesHistoryCopy
        }
    }

    val visibleStrokeSize: Int
        get() {
            return if (mVisibleStrokesHistory.value != null) {
                mVisibleStrokesHistory.value!!.size
            } else {
                -1
            }
        }

    fun getVisibleStroke(index: Int): Path? {
        return if (mVisibleStrokesHistory.value != null) {
            mVisibleStrokesHistory.value!![index]
        } else {
            null
        }
    }

    fun clearVisibleStrokes() {
        mVisibleStrokesHistory.value = ArrayList()
    }

    val currentVisibleStroke: Path?
        get() = mCurrentVisibleStroke.value

    fun setCurrentVisibleStroke(path: Path) {
        mCurrentVisibleStroke.value = path
    }

    fun resetCurrentVisibleStroke() {
        mCurrentVisibleStroke.value = Path()
    }

    fun lineToCurrentVisibleStroke(x: Float, y: Float) {
        val currentVisibleStrokeCopy = Path(mCurrentVisibleStroke.value)
        currentVisibleStrokeCopy.lineTo(x, y)
        mCurrentVisibleStroke.value = currentVisibleStrokeCopy
    }

    fun moveToCurrentVisibleStroke(x: Float, y: Float) {
        val currentVisibleStrokeCopy = Path(mCurrentVisibleStroke.value)
        currentVisibleStrokeCopy.moveTo(x, y)
        mCurrentVisibleStroke.value = currentVisibleStrokeCopy
    }

    fun setInkBuilder(mInkBuilder: Ink.Builder?) {
        this.mInkBuilder.value = mInkBuilder
    }

    fun inkBuilderBuild(): Ink? {
        return if (mInkBuilder.value != null) {
            mInkBuilder.value!!.build()
        } else {
            null
        }
    }

    fun addInkBuilderStroke(stroke: Ink.Stroke?) {
        val builder = Ink.Builder()
        val inkBuilderCopy = inkBuilderBuild()
        for (iStroke in inkBuilderCopy!!.strokes) {
            builder.addStroke(iStroke!!)
        }
        builder.addStroke(stroke!!)
        mInkBuilder.value = builder
    }

    fun removeFromStrokesHistory(index: Int) {
        if (mStrokesHistory.value != null) {
            val strokesHistoryCopy = ArrayList(mStrokesHistory.value!!)
            strokesHistoryCopy.removeAt(index)
            mStrokesHistory.value = strokesHistoryCopy
        }
    }

    val strokesHistorySize: Int
        get() {
            return if (mStrokesHistory.value != null) {
                mStrokesHistory.value!!.size
            } else {
                -1
            }
        }

    fun clearStrokesHistory() {
        mStrokesHistory.value = ArrayList()
    }

    fun addToStrokesHistory(strokeBuilder: Ink.Stroke.Builder) {
        if (mStrokesHistory.value != null) {
            val strokesHistoryCopy = ArrayList(mStrokesHistory.value!!)
            strokesHistoryCopy.add(strokeBuilder)
            mStrokesHistory.value = strokesHistoryCopy
        }
    }

    fun strokeHistoryBuild(index: Int): Ink.Stroke? {
        return if (mStrokesHistory.value != null) {
            mStrokesHistory.value!![index].build()
        } else {
            null
        }
    }

    fun getStrokeBuilder(): Ink.Stroke.Builder? {
        return mStrokeBuilder.value
    }

    fun strokeBuilderBuild(): Ink.Stroke? {
        return if (mStrokeBuilder.value != null) {
            mStrokeBuilder.value!!.build()
        } else {
            null
        }
    }

    fun addStrokeBuilderPoint(point: Ink.Point) {
        val builder = Ink.Stroke.builder()
        val strokeBuilderCopy = strokeBuilderBuild()
        for (iPoint in strokeBuilderCopy!!.points) {
            builder.addPoint(iPoint)
        }
        builder.addPoint(point)
        mStrokeBuilder.value = builder
    }

    fun clearStrokeBuilder() {
        mStrokeBuilder.value = Ink.Stroke.builder()
    }

    fun setRadicalsList(radicalsParentModels: List<Array<String>>) {
        radicalsList.postValue(radicalsParentModels)
    }
}
